### PR TITLE
Sidecar

### PR DESCRIPTION
**Pull request by simonjmendelsohn**

### CODE REVIEW

File: sfkit/sidecar/server.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+                        if not rlist:
+                            process.kill()
+
+                        for stream in rlist:
+                            line = stream.readline().strip()
+                            print(line)
+                            client.sendall(line.encode("utf-8"))
+
+                    process.wait()
```
Comment (dinvlad) [inline sfkit/sidecar/server.py:53]:
```suggestion
                    while True:
                        line = process.stdout.readline().strip()
                        if line:
                            client.sendall(line.encode("utf-8"))
                            print(line)
                        elif process.poll() is not None:
                            break
```
Simplifying, thanks to combined stderr/stdout above.

File: sfkit/sidecar/server.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            if not data:
+                break
+            try:
+                request = json.loads(data.decode("utf-8"))
+            except json.JSONDecodeError:
+                client.sendall("Invalid JSON format".encode("utf-8"))
+                continue
+
+            study_id = request.get("study_id", "")
+            data_path = request.get("data_path", "")
```
Comment (dinvlad) [inline sfkit/sidecar/server.py:24]:
```suggestion
            data_path = os.path.realpath(request.get("data_path", ""))
            
            if not data_path.startswith(constants.SAFE_DATA_PATH):
                client.sendall("Invalid data_path".encode("utf-8"))
                client.close()
                return
```
This avoids a possible path traversal attack.. We should also pair this with
```py
# constants.py

SAFE_DATA_PATH = os.environ.get("SAFE_DATA_PATH", "/data")
SAFE_DATA_PATH = os.path.join(os.path.realpath(SAFE_DATA_PATH), "")
```
(`SAFE_DATA_PATH` env var will be set to `/home/jupyter` when running as a Leo VM container)

`realpath` here resolves any symbolic links (if present), and `join` with `""` ensures there's a trailing slash, which is important to reject user-defined paths like `/home/jupyter2` later.

File: sfkit/sidecar/server.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+            client.sendall("Received request".encode("utf-8"))
+
+            command = ["env", "PYTHONUNBUFFERED=x", "sfkit", "all"]
+            if study_id:
+                command.extend(["--study_id", study_id])
+            if data_path:
+                command.extend(["--data_path", data_path])
```
Comment (dinvlad) [inline sfkit/sidecar/server.py:32]:
```suggestion
            command.extend(["--data_path", data_path])
```

**Review by dinvlad — CHANGES_REQUESTED** [review]

File: sfkit/sidecar/server.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                request = json.loads(data.decode("utf-8"))
+            except json.JSONDecodeError:
+                client.sendall("Invalid JSON format".encode("utf-8"))
+                continue
+
+            study_id = request.get("study_id", "")
+            data_path = os.path.realpath(request.get("data_path", ""))
+
+            if not data_path.startswith(constants.SAFE_DATA_PATH):
+                client.sendall("Invalid data_path".encode("utf-8"))
```
Comment (dinvlad) [inline sfkit/sidecar/server.py:26]:
```suggestion
                client.sendall(f"data_path should start with {constants.SAFE_DATA_PATH}".encode("utf-8"))
```
Might be good to give them feedback on what to expect

**Review by dinvlad — APPROVED** [review]

**Review by zbedo — APPROVED** [review]

**Pull request merged**